File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n = lists.size();
        if(n==0) return NULL;
        if(n==1) return lists.back();
        return mergesort(lists,0,n-1);
    }

    ListNode* merge(ListNode* l,ListNode* r){
        ListNode* dummy = new ListNode(-1);
        ListNode *node = dummy;
        while(l && r){
            if(l->val<r->val){
                node->next=l;
                l=l->next;
            }
            else{
                node->next=r;
                r=r->next;
            }
            node=node->next;
        } 
        if(l) node->next=l;
        if(r) node->next=r;
        return dummy->next;
    }
    ListNode* mergesort(vector<ListNode*>& lists,int s,int e){
        // if(s>e) return NULL;
        if(s==e) return lists[s];

        int mid = s+(e-s)/2;
        ListNode* l=mergesort(lists,s,mid);
        ListNode* r=mergesort(lists,mid + 1,e);

        return merge(l,r);
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        int amount = lists.length;
        int interval = 1;
        while (interval < amount) {
            for (int i = 0; i < amount - interval; i += interval * 2) {
                lists[i] = merge2Lists(lists[i], lists[i + interval]);
            }
            interval *= 2;
        }
        return amount > 0 ? lists[0] : null;
    }

    ListNode merge2Lists(ListNode l1, ListNode l2) {
        if (l1 == null || l2 == null) {
            return l1 == null? l2 : l1;
        }

        if (l1.val <= l2.val) {
            l1.next = merge2Lists(l1.next, l2);
            return l1; 
        } else {
            l2.next = merge2Lists(l1, l2.next);
            return l2;
        }
    }
}